checkers/preferDecodeRune_checker.go

```diff
@@ -0,0 +1,56 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/types"
+
+	"github.com/go-toolsmith/astcast"
+	"github.com/go-toolsmith/typep"
+
+	"github.com/go-critic/go-critic/checkers/internal/astwalk"
+	"github.com/go-critic/go-critic/framework/linter"
+)
+
+func init() {
+	var info linter.CheckerInfo
+	info.Name = "preferDecodeRune"
+	info.Tags = []string{"performance"}
+	info.Summary = "Detects expressions like []rune(s)[0] that may cause unwanted rune slice allocation"
+	info.Before = `r := []rune(s)[0]`
+	info.After = `r, _ := utf8.DecodeRuneInString(s)`
+	info.Note = `See Go issue for details: https://github.com/golang/go/issues/45260`
+
+	collection.AddChecker(&info, func(ctx *linter.CheckerContext) (linter.FileWalker, error) {
+		return astwalk.WalkerForExpr(&preferDecodeRuneChecker{ctx: ctx}), nil
+	})
+}
+
+type preferDecodeRuneChecker struct {
+	astwalk.WalkHandler
+	ctx *linter.CheckerContext
+}
+
+func (c *preferDecodeRuneChecker) VisitExpr(e ast.Expr) {
+	indexExpr := astcast.ToIndexExpr(e)
+	indexed := indexExpr.X
```

Comment:
This can be `nil` afair. Will next line fail? 👀

Reply:
AFAIK, `TypeOf(nil)` is safe.
And astcast returns a sentinel object in case of a failed case, so it's safe to access its fields.